File: packages/python-mdi-llmkit/src/mdi_llmkit/gpt_api/json_schema_format.py
from typing import Any, Dict, List, Optional, Union


def JSONSchemaFormat(schema: Any, *, name: str, description: str):
    """A convenience function that allows us to easily create JSON schema formats."""
    retval = {
        "format": {
            "type": "json_schema",
            "strict": True,
        },
    }
    if name:
        retval["format"]["name"] = name
    if description:
        retval["format"]["description"] = description

    TYPEMAP = {
        str: "string",
        int: "integer",
        float: "number",
        bool: "boolean",
    }

    def _convert_schema_recursive(subschema: Any) -> dict:
        # If the subschema is a Tuple, then it will consist of either two or three elements.
        # One of these elements will be a string. The other element will be a list of strings.
        # The third element will be the subschema's value. These elements can occur in any order.
        # Oh, it can also be a pair of numerical values, which represent min and max ranges.
        subschema_description = ""
        subschema_enum = []
        subschema_numrange = (None, None)
        subschema_value = subschema
        if isinstance(subschema, tuple):
            for item in subschema:
                if not item:
                    # If the item is falsy, then it's a data type placeholder.
                    subschema_value = item
                    continue

                # If it's a string and it wasn't falsy, then assume it's a description.
                if isinstance(item, str):
                    subschema_description = item
                    continue

                # If it's a list of length >= 2 and all list members are strings, then assume it's an enum.
                if (
                    isinstance(item, list)
                    and len(item) >= 2
                    and all(isinstance(i, str) for i in item)
                ):
                    subschema_enum = item
                    continue

                # If it's a tuple of length 2 and at least one element is a float or int,
                # then assume it's a numeric range.
                if (
                    isinstance(item, tuple)
                    and len(item) == 2
                    and (
                        isinstance(item[0], (float, int))
                        or isinstance(item[1], (float, int))
                    )
                ):
                    subschema_numrange = item
                    continue

                # At this point, we have to assume that the item is the schema value.
                subschema_value = item

        if isinstance(subschema_value, tuple):
            # We might be able to infer its type by its enum or range.
            if len(subschema_enum) > 0:
                # It's implicitly a string.
                subschema_value = str

            nr0 = subschema_numrange[0]
            nr1 = subschema_numrange[1]
            if nr0 is not None or nr1 is not None:
                if isinstance(nr0, float) or isinstance(nr1, float):
                    subschema_value = float
                else:
                    subschema_value = int

        recretval = {}

        if isinstance(subschema_value, dict):
            recretval["type"] = "object"
            if subschema_description:
                recretval["description"] = subschema_description
            recretval["additionalProperties"] = False
            recretval["required"] = [p for p in subschema_value.keys()]
            recretval["properties"] = {}
            for k, v in subschema_value.items():
                if isinstance(v, str):
                    recretval["properties"][k] = {"type": "string", "description": v}
                else:
                    recretval["properties"][k] = _convert_schema_recursive(v)

        elif isinstance(subschema_value, list):
            # If it's a list of length >= 2 and all list members are strings, then assume it's an enum.
            if len(subschema_value) >= 2 and all(
                isinstance(i, str) for i in subschema_value
            ):
                recretval["type"] = "string"
                subschema_enum = subschema_value
            else:
                recretval["type"] = "array"
                if subschema_description:
                    recretval["description"] = subschema_description
                if subschema_numrange[0] is not None:
                    recretval["minItems"] = subschema_numrange[0]
                if subschema_numrange[1] is not None:
                    recretval["maxItems"] = subschema_numrange[1]
                arrayexemplar = subschema_value[0]
                if isinstance(arrayexemplar, str):
                    recretval["items"] = {
                        "type": "string",
                        "description": arrayexemplar,
                    }
                else:
                    recretval["items"] = _convert_schema_recursive(arrayexemplar)

        else:
            subschema_type = TYPEMAP.get(subschema_value)
            if not subschema_type:
                subschema_type = TYPEMAP.get(type(subschema_value))
            if not subschema_type:
                raise ValueError(
                    f"Unrecognized type for schema value: {subschema_value}"
                )
            recretval["type"] = subschema_type
            if subschema_description:
                recretval["description"] = subschema_description

        if subschema_enum:
            recretval["enum"] = subschema_enum

        if isinstance(subschema_value, (int, float)):
            if subschema_numrange[0] is not None:
                recretval["minValue"] = subschema_numrange[0]
            if subschema_numrange[1] is not None:
                recretval["maxValue"] = subschema_numrange[1]

        return recretval

    convresult = _convert_schema_recursive(schema)
    if convresult["type"] != "object":
        if not name:
            name = "schema"
        convresult = {
            "type": "object",
            "required": [name],
            "additionalProperties": False,
            "properties": {name: convresult},
        }

    retval["format"]["schema"] = convresult
    return retval

File: packages/typescript-mdi-llmkit/src/gptApi/jsonSchemaFormat.ts
export const JSON_INTEGER = Symbol('JSON_INTEGER');
export const JSON_NUMBER = Symbol('JSON_NUMBER');
export const JSON_STRING = String;
export const JSON_BOOLEAN = Boolean;

export interface JSONSchemaFormatResult extends Record<string, unknown> {
  format: {
    type: 'json_schema';
    strict: true;
    name?: string;
    description?: string;
    schema: Record<string, unknown>;
  };
}

const TYPEMAP = new Map<unknown, string>([
  [JSON_STRING, 'string'],
  [JSON_INTEGER, 'integer'],
  [JSON_NUMBER, 'number'],
  [JSON_BOOLEAN, 'boolean'],
  [String, 'string'],
  [Boolean, 'boolean'],
  [BigInt, 'integer'],
  [Number, 'number'],
]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function isStringArray(value: unknown): value is string[] {
  return (
    Array.isArray(value) && value.every((item) => typeof item === 'string')
  );
}

function isNumericRangeArray(
  value: unknown
): value is [number | null, number | null] {
  if (!Array.isArray(value) || value.length !== 2) {
    return false;
  }
  const [min, max] = value;
  const minValid = min === null || typeof min === 'number';
  const maxValid = max === null || typeof max === 'number';

  return (
    minValid && maxValid && (typeof min === 'number' || typeof max === 'number')
  );
}

function isTupleMetadataArray(value: unknown): value is unknown[] {
  if (!Array.isArray(value) || value.length < 2) {
    return false;
  }
  if (isStringArray(value)) {
    return false;
  }

  return value.some(
    (item) => typeof item === 'string' || isNumericRangeArray(item)
  );
}

function inferPrimitiveType(schemaValue: unknown): string | null {
  const direct = TYPEMAP.get(schemaValue);
  if (direct) {
    return direct;
  }

  if (typeof schemaValue === 'string') {
    return 'string';
  }
  if (typeof schemaValue === 'boolean') {
    return 'boolean';
  }
  if (typeof schemaValue === 'bigint') {
    return 'integer';
  }
  if (typeof schemaValue === 'number') {
    return Number.isInteger(schemaValue) ? 'integer' : 'number';
  }

  return null;
}

function convertSchemaRecursive(subschema: unknown): Record<string, unknown> {
  let subschemaDescription = '';
  let subschemaEnum: string[] = [];
  let subschemaNumrange: [number | null, number | null] = [null, null];
  let subschemaValue: unknown = subschema;

  if (isTupleMetadataArray(subschema)) {
    for (const item of subschema) {
      if (!item) {
        subschemaValue = item;
        continue;
      }

      if (typeof item === 'string') {
        subschemaDescription = item;
        continue;
      }

      if (isStringArray(item) && item.length >= 2) {
        subschemaEnum = item;
        continue;
      }

      if (isNumericRangeArray(item)) {
        subschemaNumrange = item;
        continue;
      }

      subschemaValue = item;
    }
  }

  if (
    (Array.isArray(subschemaValue) && isTupleMetadataArray(subschemaValue)) ||
    (Array.isArray(subschemaValue) && subschemaValue.length === 0)
  ) {
    if (subschemaEnum.length > 0) {
      subschemaValue = JSON_STRING;
    }

    const [nr0, nr1] = subschemaNumrange;
    if (nr0 !== null || nr1 !== null) {
      if (
        (typeof nr0 === 'number' && !Number.isInteger(nr0)) ||
        (typeof nr1 === 'number' && !Number.isInteger(nr1))
      ) {
        subschemaValue = JSON_NUMBER;
      } else {
        subschemaValue = JSON_INTEGER;
      }
    }
  }

  const result: Record<string, unknown> = {};

  if (isRecord(subschemaValue)) {
    result.type = 'object';
    if (subschemaDescription) {
      result.description = subschemaDescription;
    }
    result.additionalProperties = false;

    const keys = Object.keys(subschemaValue);
    result.required = keys;

    const properties: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(subschemaValue)) {
      if (typeof value === 'string') {
        properties[key] = { type: 'string', description: value };
      } else {
        properties[key] = convertSchemaRecursive(value);
      }
    }

    result.properties = properties;
  } else if (Array.isArray(subschemaValue)) {
    if (subschemaValue.length >= 2 && isStringArray(subschemaValue)) {
      result.type = 'string';
      subschemaEnum = subschemaValue;
    } else {
      result.type = 'array';
      if (subschemaDescription) {
        result.description = subschemaDescription;
      }
      if (subschemaNumrange[0] !== null) {
        result.minItems = subschemaNumrange[0];
      }
      if (subschemaNumrange[1] !== null) {
        result.maxItems = subschemaNumrange[1];
      }

      const arrayExemplar = subschemaValue[0];
      if (typeof arrayExemplar === 'string') {
        result.items = { type: 'string', description: arrayExemplar };
      } else {
        result.items = convertSchemaRecursive(arrayExemplar);
      }
    }
  } else {
    const primitiveType = inferPrimitiveType(subschemaValue);
    if (!primitiveType) {
      throw new Error(
        `Unrecognized type for schema value: ${String(subschemaValue)}`
      );
    }
    result.type = primitiveType;
    if (subschemaDescription) {
      result.description = subschemaDescription;
    }
  }

  if (subschemaEnum.length) {
    result.enum = subschemaEnum;
  }

  if (result.type === 'integer' || result.type === 'number') {
    if (subschemaNumrange[0] !== null) {
      result.minimum = subschemaNumrange[0];
    }
    if (subschemaNumrange[1] !== null) {
      result.maximum = subschemaNumrange[1];
    }
  }

  return result;
}

export function JSONSchemaFormat(
  name: string,
  schema: unknown,
  description?: string
): JSONSchemaFormatResult {
  const result: JSONSchemaFormatResult = {
    format: {
      type: 'json_schema',
      strict: true,
      name,
      schema: {
        type: 'object',
        properties: {},
        required: [],
        additionalProperties: false,
      },
    },
  };

  if (description) {
    result.format.description = description;
  }

  let converted = convertSchemaRecursive(schema);
  if (converted.type !== 'object') {
    converted = {
      type: 'object',
      required: [name],
      additionalProperties: false,
      properties: {
        [name]: converted,
      },
    };
  }

  result.format.schema = converted;
  return result;
}
